src/main/java/net/schmizz/sshj/common/KeyType.java

```diff
@@ -372,8 +396,14 @@ private static Date dateFromEpoch(BigInteger seconds) {
             }
         }
 
-        private static long epochFromDate(Date date) {
-            return date.getTime() / 1000;
+        private static BigInteger epochFromDate(Date date) {
+            long time = date.getTime();
+            if (time >= Long.MAX_VALUE / 1000 * 1000) {
+                // OpenSSH expects this number when the date is the infinite future.
+                return Buffer.MAX_UINT64_VALUE;
```

Comment:
No, it isn't. `long` in java is always signed, and the last bit holds the sign. `Long.MAX_VALUE` is 2⁶³ - 1, while `Buffer.MAX_UINT_VALUE` is 2⁶⁴ - 1.
It still can be repesented as long as `0xFFFFFFFFFFFFFFFFL`, it's absolutely the same as `-1L`. However, I see `Buffer.MAX_UINT64_VALUE` explaining itself better than the combination of `-1` and `putUInt64Unchecked`.

Reply:
Also, if you want to get rid of juggling with `long` and `BigInteger`, we have to get rid of `Date` in favour of `BigInteger`, because `Date` is strictly bound to the signed `long`.

Reply:
That might be a better idea indeed. The code is somewhat convoluted, and non self-explanatory now...